input_algorithms/validators.py

```diff
@@ -65,6 +65,28 @@ def validate(self, meta, val):
         return val
 
 @register
+class has_only_one_of(Validator):
+    """
+    Usage
+        .. code-block:: python
+
+            has_only_one_of([key1, ..., keyn]).normalise(meta, val)
+
+    Will complain if the ``val.get(key, NotSpecified)`` returns ``NotSpecified``
+    for all but one of the choices.
+
+    I.e. A valid dictionary must have exactly one of the specified keys!
+    """
+    def setup(self, choices):
+        self.choices = choices
```

Comment:
We should check there is atleast one choice specified and raise input_algorithms.errors.ProgrammerError if they have not.